sympy/core/exprtools.py

```diff
@@ -1414,156 +1288,614 @@ def factor_nc(expr):
-
-        new_mid = new_mid.subs(r2).subs(unrep1)
-
-        if new_mid.is_Pow:
-            return _keep_coeff(c, g*l*new_mid*r)
-
-        if new_mid.is_Mul:
-            def _pemexpand(expr):
-                "Expand with the minimal set of hints necessary to check the result."
-                return expr.expand(deep=True, mul=True, power_exp=True,
-                    power_base=False, basic=False, multinomial=True, log=False)
-            # XXX TODO there should be a way to inspect what order the terms
-            # must be in and just select the plausible ordering without
-            # checking permutations
-            cfac = []
-            ncfac = []
-            for f in new_mid.args:
-                if f.is_commutative:
-                    cfac.append(f)
-                else:
-                    b, e = f.as_base_exp()
-                    if e.is_Integer:
-                        ncfac.extend([b]*e)
-                    else:
-                        ncfac.append(f)
-            pre_mid = g*Mul(*cfac)*l
-            target = _pemexpand(expr/c)
-            for s in variations(ncfac, len(ncfac)):
-                ok = pre_mid*Mul(*s)*r
-                if _pemexpand(ok) == target:
-                    return _keep_coeff(c, ok)
-
-        # mid was an Add that didn't factor successfully
-        return _keep_coeff(c, g*l*mid*r)
+            if not ok:
+                continue
+
+            substitutes = list(zip(dummies, soln))
+            yield (
+                _NCPoly(Gexpr.subs(substitutes), self.generators, self.extensions),
+                _NCPoly(Hexpr.subs(substitutes), self.generators, self.extensions),
+            )
+
+    def _factor_recursive(self):
+        """Return a generator that yields all unique, valid, irreducible factorizations.
+        Does not fully simplify coefficients.
+        """
+        self_expr = self.as_expr()
+        out = set()
+        for degree in range(1, self.degree):
+            for G, H in self._factor_once(degree):
+                Gcoeff, G = G.factor_coeff()
+                Hcoeff, H = H.factor_coeff()
+                coeff = Gcoeff * Hcoeff
+                Gs = G._factor_recursive()
+                Hs = H._factor_recursive()
+                H_list = []
+                G_first = next(Gs)
+                coeff_G = coeff * G_first
+                for H_expr in Hs:
+                    H_list.append(H_expr)
+                    expr = coeff_G * H_expr
+                    if expr not in out:
+                        yield expr
+                        out.add(expr)
+                for G_expr in Gs:
+                    coeff_G = coeff * G_expr
+                    for H_expr in H_list:
+                        expr = coeff_G * H_expr
+                        if expr not in out:
+                            yield expr
+                            out.add(expr)
+        if len(out) == 0:
+            yield self_expr
+
+    def factor_one(self):
```

Comment:
I would also not tie `factor` to the class at begin. 
Maybe only four arithmetic and a couple of friends could be part of `NCPoly`, because they can be friends, but I don't think that `factor` can be friend to very rudimentary arithmetic.